implemented the merch store

sumeet-bansal opened the pull request:
Closes #53.

shravanhariharan2 commented on src/email/templates/orderConfirmation.ejs:
```diff
@@ -0,0 +1,166 @@
+                                                            <th>Quantity</th>
+                                                            <th>Total</th>
+                                                        </tr>
+
+                                                        <% for(var i=0; i < order.items.length; i++) { %>
+                                                            <tr>
+                                                                <td><%= order.items[i].itemName %></td>
+                                                                <td><%= order.items[i].price %></td>
+                                                                <td><%= order.items[i].quantityRequested %></td>
+                                                                <td><%= order.items[i].total %></td>
+                                                            </tr>
+                                                        <% } %>
```
The `price` and `total` don't reflect discounted items, they reflect the actual price of the items even though the discounted price is subtracted from the user's credits. I think you should probably display discounted price here rather than raw price

sumeet-bansal replied:
Good catch, will do.

sumeet-bansal commented:
> If two or more errors are encountered, I think it would be useful for users to see all errors that they encountered, such as "error: monthly limit reached for x, lifetime limit reached for y, out of stock for z" instead of just listing "monthly limit reached for x" because of its priority. This way, users ordering multiple items can fix all their ordering issues at once instead of fixing each one individually and running into the next one when ordering again.

> Then again, I feel like it would be beneficial to warn the user of these errors before they even select the erroneous items and try to place their order (like in a shopping list in the FE, gray-out the "Add item" button if monthly/lifetime limits are reached). Although, this would probably result in a bunch of extra API calls to load all that info, so the current solution may be the most performant.

The FE could preload the user's purchase history (`GET /store/order`) and use that to determine if an order exceeds some monthly/lifetime limit, and the FE already has all the data it needs (credits, merch prices and discounts) to determine if the user has enough credits, so those are UX improvements that the FE can implement without any further changes to the store controller.

> Are discounts permanent/static? Or do they last a certain amount days and have to be manually updated by an admin once to start/end the discount? Since there was no timestamp attached to a Merchandise item, I'm assuming its static but I think it would be cool to have temporary discounts for items, perhaps for celebration of some event/activity (like the Camp Snoopy Snapback going on sale for 1 week to celebrate some Sixth College event)

Discounts can be updated, and I debated keeping track of price/discount changes or at what price an item gets bought, so `Order.totalCost` kind of logs that (and there could be an associated `OrderItem.cost`), but I didn't really put a lot of thought into it so I'm open to suggestions. Maybe `OrderItem.cost` and `OrderItem.discount`? So we can determine the original price, discount, and discounted price at which any item is purchased at.

> Is there any functional/monetary benefit of collections besides the fact that they are a group of similar items? I saw you mentioned perhaps having discounts over collections, and I think that would be super cool (like Hack School collection having all items 20% off for a week to celebrate the last week of Hack School).

In our initial store planning meetings, we wanted to do limited "drops" and collections provide a way to logically group together similar items. Collections as a whole can be discounted, archived, etc., so it's easier for admins to make changes to an entire group of merch items as well (although the PR doesn't currently have a way for collections to be discounted or archived&mdash;I'll add that). I'm thinking some collection-wide operations just write to all the collection's items instead of adding attributes to `Collection` itself (e.g. changing discount percentage), `Collections` gets an `archived` attribute, and we delete `GET /store/merchandise` since it basically does the same thing as `GET /store/collection` but without the collection info.

shravanhariharan2 commented on src/api/v1/store/index.js:
```diff
@@ -0,0 +1,337 @@
+      }
+      const findCollectionsMinusMerch = req.user.isAdmin()
+        ? MerchandiseCollection.getAllCollections() : MerchandiseCollection.getAllUnarchivedCollections();
+      const collectionsMinusMerch = (await findCollectionsMinusMerch).then((cs) => cs.map((c) => c.get()));
```
Does this line cause an error for you? I'm getting a `TypeError: (intermediate value).then is not a function` on the `.then` when calling GET `/api/v1/store/collection/` as an admin (errors out as non-admin as well). I'm on the latest pull from this branch too.

sumeet-bansal replied:
Fixed.

steets250 commented on src/db/models/merchandise.js:
```diff
@@ -0,0 +1,183 @@
+  Merchandise.getDiscountedItems = function (offset, limit) {
+    if (!offset || offset < 0) offset = 0;
+    if (!limit || limit < 0) limit = undefined;
+    return this.findAll({
```
Do you also need to check for `hidden: false` here for non-admin users?

sumeet-bansal replied:
The check could be elsewhere, e.g. the API layer, but I don't think this function is being used so I'm just gonna delete it.

shravanhariharan2 commented on src/api/v1/store/index.js:
```diff
@@ -0,0 +1,337 @@
+    return next();
+  })
+
+  /**
+   * Creates an item, given a 'merchandise' object in the request body, and returns the created item
+   * upon success. Required fields: itemName, price, description. Optional fields: picture, quantity,
+   * discountPercentage, hidden. All other fields will be ignored.
+   */
+  .post(async (req, res, next) => {
+    if (!req.body.merchandise) return next(new error.BadRequest('Merchandise item must be provided'));
+
```
POST `/store/merchandise/` throws an error if no collection uuid is provided or if collection is null in request body. Not all items need to belong to a collection though, right?

sumeet-bansal replied:
Items must belong to a collection; it makes implementation (and merch taxonomy) much simpler and design team signed off on it.

sumeet-bansal commented:
> Email confirmation: email still doesn't show the proper discounted price after ordering. When I buy a 2000 credit item for 20% off, 1600 credits gets taken away from my acc but email confirmation shows 2000. Also the email CSS is still a little funky but I remember you mentioning someone else was looking into that already.

Correct sale price gets passed into the email template now.

> `GET /order`: I'm getting an empty response even after ordering items. As both an admin and a user, I get no items at all even after I order items and receive confirmation.

Fixed.

> `POST /order`: This route doesn't work when ordering more than 1 item now. I get a BadRequest [400]: There are duplicate or invalid items in this order when sending an order with more than 1 item.

Fixed.

> `PATCH /order`: I'm not sure this route properly fulfills the item. Boolean value fulfilled still remains "f" for OrderItems table when looking at database after sending those uuids into the request. This is after ordering the items I mentioned above as separate orders.
> `PATCH /order`: Doesn't reject my error when I had one item "f" and one item "t" and sent both to fulfill. I didn't get an error, but nothing changed in the database regardless.

Works as intended. Each UUID has a couple fields to go with it: `fulfilled: boolean` and `notes: string`, so a distributor can add a note to an OrderItem without fulfilling it.

shravanhariharan2 commented on src/email/templates/orderConfirmation.ejs:
```diff
@@ -0,0 +1,132 @@
+<table border="0" cellpadding="0" cellspacing="0" class="body" style="border-collapse: separate; mso-table-lspace: 0pt; mso-table-rspace: 0pt; width: 100%; background-color: #f6f6f6;">
+    <tr>
+        <td style="font-family: sans-serif; font-size: 14px; vertical-align: top;">&nbsp;</td>
+        <td class="container" style="font-family: sans-serif; font-size: 14px; vertical-align: top; display: block; Margin: 0 auto; max-width: 580px; padding: 10px; width: 580px;">
+            <div class="content" style="box-sizing: border-box; display: block; Margin: 0 auto; max-width: 580px; padding: 10px;">
+
+<% for(var i=0; i < order.items.length; i++) { %>
+                    <p><%= order.items[i].itemName %></p>
+                    <p><small><%= order.items[i].salePrice %> x <%= order.items[i].quantityRequested %> = <%= order.items[i].total %></small></p>
+<% } %>
+    <h1><%= order.items[i].totalCost %></h1>
+                                        <p style="font-family: sans-serif; font-size: 14px; font-weight: normal; margin: 0; Margin-bottom: 15px;">To pick up your items, find a merch distribution event on the membership portal and stop by.</p>
+                                        <p style="font-family: sans-serif; font-size: 14px; font-weight: normal; margin: 0; Margin-bottom: 15px;">Regards, <br />ACM UCSD</p>
+                                    </td>
+                                </tr>
+                            </table>
```
Getting an error caused by these lines when ordering any item. The order still goes through but I don't get any confirmation email.
```
InternalServerError: ejs:100
    98|                     <p><small><%= order.items[i].salePrice %> x <%= order.items[i].quantityRequested %> = <%= order.items[i].total %></small></p>
    99| <% } %>
 >> 100|     <h1><%= order.items[i].totalCost %></h1>
    101|                                         <p style="font-family: sans-serif; font-size: 14px; font-weight: normal; margin: 0; Margin-bottom: 15px;">To pick up your items, find a merch distribution event on the membership portal and stop by.</p>
    102|                                         <p style="font-family: sans-serif; font-size: 14px; font-weight: normal; margin: 0; Margin-bottom: 15px;">Regards, <br />ACM UCSD</p>
    103|                                     </td>

Cannot read property 'totalCost' of undefined
```

Input request:
```
{
	"order": [
		{
			"item": "6b65bec1-4bff-4334-bf64-792b83ed426a",
			"quantity": 1
		}
	]
}
```
`order` response when console.logged:
```
{
  uuid: '1ea1fbee-7aa4-4257-b869-d3a7173e51ce',
  orderedAt: 2020-05-20T18:56:15.360Z,
  totalCost: 1275,
  items: [
    {
      itemName: 'Hack School Sticker Pack (4)',
      picture: null,
      description: 'Make space on your laptop cover for these 2 in x 2 in stickers. Pack of 4.',
      quantityRequested: 1,
      salePrice: 1275,
      total: 1275
    }
  ]
```

sumeet-bansal replied:
Fixed typo.

sumeet-bansal commented:
Oops, must've deleted some HTML by accident. Added it back.
